types: use enum class for enums

perrinjerome opened the pull request:
## Description 

Using enums allows more precise type checking. The current code in types.py was already using *Kind classes as if they were enums, which mypy reported as type errors:
    
    pygls/types.py:377: error: Incompatible default for argument "severity" (default has type "int", argument has type "DiagnosticSeverity")
    pygls/types.py:663: error: Incompatible default for argument "trace" (default has type "str", argument has type "Optional[Trace]")

WatchKind is handled a bit differently, because valid values can be combined with |, like the default value 7.

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

danixeee commented on pygls/protocol.py:
```diff
@@ -369,8 +370,12 @@ def _send_data(self, data):
         if not data:
             return
 
+        def default(o):
```
Could you just move this function under `call_user_feature`(line 70.) and perhaps rename it to `default_serializer` or something similar?

perrinjerome replied:
Thanks, good idea, I pushed more fixup commits to address this.

danixeee commented:
@perrinjerome PR https://github.com/openlawlibrary/pygls/pull/90 is merged. If you feel comfortable, you can rebase branch with master and update the changelog for this PR.

perrinjerome commented:
Thanks @danixeee . I was not familiar with `enum` module, and after doing a bit more research, I realised that using `enum.Enum` was incorrect, because protocol is not deserialised as actual objects from `pygls.types` module.

For example when language server is implemented as:

```python
@server.feature(COMPLETION, trigger_characters=[','])
def completions(params: CompletionParams = None) -> CompletionList:
    """Returns completion items."""
    return CompletionList(False, [])
```

what this function receive as `params` argument is not a `CompletionParams`, but a "simple" `NamedTuple`, as protocol messages are deserialized here:

https://github.com/openlawlibrary/pygls/blob/df8950dac6539e387817d3f310c18465d8997204/pygls/protocol.py#L83

This also means that `params.context.triggerKind`, which is supposed to be of type `TriggerKind` enum is not an enum but a plain `int` and tests such as ( this comes from [cmake-language-server](https://github.com/regen100/cmake-language-server/blob/66af586b2aa4da1a21b54e566f81d250acb0a848/src/cmake_language_server/server.py#L43-L46) )

```python
        @self.feature(COMPLETION, trigger_characters=['{', '('])
        def completions(params: CompletionParams):
            if (params.context.triggerKind ==
                    CompletionTriggerKind.TriggerCharacter):
```
will be always false, because int are different from enum.

This small issue also interesting in context of https://github.com/openlawlibrary/pygls/issues/81 


But the problem seems that I wrongly used `enum.Enum` in this patch. When using `enum.IntEnum`, everything works fine  ( reference doc: https://docs.python.org/3/library/enum.html#intenum ) :

```
>>> import enum
>>> class wrong(enum.Enum):
...   one = 1
... 
>>> class ok(enum.IntEnum):
...   one = 1
... 
>>> 1 == wrong.one
False
>>> 1 == ok.one
True
```

So I will amend this patch to use `IntEnum` for int.

I will also change the string enums to inherit from `str` as suggested in https://docs.python.org/3/library/enum.html#others 

I will also revisit the `WatchKind` case to use an [`enum.IntFlag`]( https://docs.python.org/3/library/enum.html#intflag).

perrinjerome commented:
> I will also revisit the `WatchKind` case to use an [`enum.IntFlag`](https://docs.python.org/3/library/enum.html#intflag).

That's what I did as part of https://github.com/openlawlibrary/pygls/pull/92/commits/c0623414c9e2e956298dc6f2568ef3da042f5977  but `IntFlag` is *New in version 3.6.*, so that's was not OK.
I'll revisit this, hopefully later this week.

perrinjerome commented:
I amended to support python 3.5 , the change compared to previous version was https://github.com/openlawlibrary/pygls/commit/354ff54e4edbe835b5dc3bcd328fd8500fce1526 .

Then I rebased on master, squashed all the intermediate "fixup" commits from this branch, added an entry in change log, edited commit message to explain the need for `enum.IntEnum` and `str` subclassing and push-force the result in this branch.

Please take another look, I think this is OK now.

perrinjerome commented:
Thank you @danixeee . I also learned about enums in the process :) Don't worry about the delay, everybody is busy.
